Parse/Parse/PFObject.m

```diff
@@ -207,7 +207,7 @@ + (BFTask *)_enqueue:(BFTask *(^)(BFTask *toAwait))taskStart forObjects:(NSArray
 + (BOOL)collectDirtyChildren:(id)node
                     children:(NSMutableSet *)dirtyChildren
                        files:(NSMutableSet *)dirtyFiles
-                        seen:(NSSet *)seen
+                        seen:(NSMutableSet *)seen
```

Comment:
Thinking about this more, this may or not be the right solution. This is a recursive call, depending on how it's used this may or may not need to be mutable.

Does the original implementation work fine on a number of smaller children? If it does, I'm assuming it's less than 50, and the issue comes up over 50. If so, the problem might be the batching calculation I mentioned.

When I wrote those test cases, when it went over the set amount to batch (50), it returned only the first 50 objects every time which sounds like a similar symptom that you are reporting.

Reply:
I don't think this is a batching issue as batching will happen after all dirty children are calculated. In my situation this recursive method spins without finishing, consuming all CPU. It does indeed work fine for small objects but explodes quickly. There is no batching involved in my case as I know there is only one dirty object, the root object, but this method will run into a seemingly endless loop trying to traverse all child objects. 

This recursive call is internal and only called in one place passing in a `NSMutableSet` (changing this to require a mutable set only failed in this one call site). 

The problem here is triggered by our complex schema with a couple of classes, where each unfortunately is linked back to a root parent class. Now when we use arrays that are traversed and at the end many pointers point to the same parent, this parent is visited many, many, many times, and this explodes very quickly.

```
p = [Parent]
a.parent = p
b.parent = p
c.parent = p
d.parent = p
r = [Root]
r.objects = [a, b, c, d]
// saving this root will visit each of the children a, b, c, d, 
// but the algorithm will not remember that p has already been visited and p will be visited 4 times,
// including all of p's children. 
r.saveInBackground()
```

This explodes exponentially when you have layers of nested pointers to the common object or if you imagine the parent having pointers to other objects. This happens before the dirty children are returned, so before any batching could occur, so I don't think https://github.com/parse-community/Parse-SDK-iOS-OSX/issues/1588 is necessarily related.

In my case this method never returns in reasonable time (I think it will return eventually, I cannot detect an endless loop in there, but that may take years).

Reply:
> p = [Parent]
a.parent = p
b.parent = p
c.parent = p
d.parent = p
r = [Root]
r.objects = [a, b, c, d]
// saving this root will visit each of the children a, b, c, d, 
// but the algorithm will not remember that p has already been visited and p will be visited 4 times,
// including all of p's children. 
r.saveInBackground()

I agree that it seems like it should remember `p` after visiting `a`.

Reply:
It also seems like this would be remedied if `p` points to all it's children and if the children really want to know their relation to their parent `p`, they just add a field referencing `p`'s `objectId`, not pointing back up to `p`, allowing everything to traverse. This was the issue that was coming up in the other issue you referenced.